keps/sig-network/1672-tracking-terminating-endpoints/README.md

```diff
@@ -148,9 +155,123 @@ of the [EndpointSlice API work](/keps/sig-network/20190603-endpointslices/README
 Since this is an addition to the EndpointSlice API, the version skew strategy will follow that
 of the [EndpointSlice API work](/keps/sig-network/20190603-endpointslices/README.md).
 
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: EndpointSliceTerminatingCondition
+  - Components depending on the feature gate: kube-apiserver and kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+
+Yes, terminating endpoints are now included as part of EndpointSlice API. The `ready` condition of an EndpointSlice will always be `false` to ensure consumers do not send traffic to terminating endpoints unless new conditions are checked.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. On rollback, terminating endpoints will no longer be included in EndpointSlice and the `terminating` and `serving` conditions will not be set.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+EndpointSlice will continue to have the `terminating` and `serving` condition set.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, there will be integration and e2e tests validating whether EndpointSlice contains endpoints for pods that are terminating.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout fail? Can it impact already running workloads?
+
+If there are consumers of EndpointSlice that do not check the `ready` condition, then they may unexpectedly start sending traffic to terminating endpoints.
+It is assumed that almost all consumers of EndpointSlice check the `ready` condition prior to allowing traffic to a pod.
+
+###### What specific metrics should inform a rollback?
+
+Application-level traffic indicating packet-loss or error rates.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Not yet, but manual upgrade and rollback testing will be done prior to graduating the feature to Beta.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The condition will always be set for terminating pods but consumers may choose to ignore them. It is up to consumers of the API to provide metrics
+on how the new conditions are being used.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+N/A
```

Comment:
I don't agree it's n/a - visibility into a feature is always applicable.

As an example, I can imagine that a useful metric would be to split the existing:
endpoints_added_per_sync
endpoints_removed_per_sync
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/metrics/metrics.go#L32

metrics, to be split be terminating condition

Reply:
I'm not sure I agree.. that is exposing API-level semantics into the metrics, are we sure we want that? You could say the same thing for every possible condition and parameter included in EndpointSlice, that would explode metrics cardinality.

Reply:
I'm fine with other metric - but I really like to see *some* metric.

Reply:
what about?
```
EndpointsDesired.WithLabelValues("Ready").Set(float64(c.numEndpointsReady))
EndpointsDesired.WithLabelValues("Serving").Set(float64(c.numEndpointsServing))
EndpointsDesired.WithLabelValues("Terminating").Set(float64(c.numEndpointsTerminating))
```

Reply:
That covers total endpoints across all EndpointSlices, that seems reasonable to me. Thoughts @wojtek-t ?

Reply:
It's hard to define "working" for this overall feature, but I agree it's interesting to think about.

First there's the split between what the controller does and what kube-proxy does.  These are two different KEPs, so I guess we should be clear which is which, but they should be coordinated.

I admit some ignorance of metrics compat - is it allowable to take a metric like `EndpointsDesired` which has no labels, and add new labels?

While yes, this is exposing some of the API thru metrics, I think it is interesting to think about "how would I know if the number of terminating endpoints was spiking?"  Breaking the metric up this way seems reasonable.  At least we can see SOMETHING

Looking at the metrics around EPSlice, I am a little suspicious of their utility.  Do I really care about the total number of endpoints in the cluster?  What questions can I answer with that?  Do we need to add metrics that give a more scaled POV?  Like the 50p/90p/99p number of terminating endpoints per service or something?

Reply:
@robscott

Reply:
>I admit some ignorance of metrics compat - is it allowable to take a metric like EndpointsDesired which has no labels, and add new labels?

Not an expert here, but it seems like this should be allowed since the same overall metric would still be available, this would be an additive change. For better or worse, it looks like we've never graduated the stability of these metrics, so they're all still listed as [alpha](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/metrics/metrics.go#L37).

> Looking at the metrics around EPSlice, I am a little suspicious of their utility.

Yeah, most of them have minimal utility. They were added when one of the most interesting questions was how efficient the slicing algorithm and associated distribution would be, but they're not great at answering if the system as a whole is working. On the other hand, I'm not really sure what kinds of metrics would help with that. 

As part of the topology KEP, I [added a metric](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/metrics/metrics.go#L110) that tracks the results of syncs (grouped by success or failure), that seems like one of the most helpful metrics we have for this controller now.

> Do we need to add metrics that give a more scaled POV? Like the 50p/90p/99p number of terminating endpoints per service or something?

I'd looked into something similar before, but I couldn't find a great way to actually implement that. Assuming we'd need to have a histogram, but not sure how we'd populate it. If we only wrote every time a Service was synced and only wrote  a value for the synced Service, I don't think we'd end up with the data we wanted here. Maybe we'd need to write the data for all Services on some kind of relatively frequent basis that was disconnected from the sync loop.

Reply:
In this example wouldn't we end up double counting endpoints?
```
EndpointsDesired.WithLabelValues("Ready").Set(float64(c.numEndpointsReady))
EndpointsDesired.WithLabelValues("Serving").Set(float64(c.numEndpointsServing))
EndpointsDesired.WithLabelValues("Terminating").Set(float64(c.numEndpointsTerminating))
```

My limited understanding of how these metrics work is that the overall value here would be a sum of the values with labels. In this case, I think it's likely that 2 of these conditions/labels would be true for a large chunk of endpoints, which could result in doubling the overall value.

Reply:
>> Do we need to add metrics that give a more scaled POV? Like the 50p/90p/99p number of terminating endpoints per service or something?
>
> I couldn't find a great way to actually implement that. 

I agree that it's not obvious.  It's kind of a distribution of distributions.  But could we do something simpler and more hardcoded?

E.g. on every EPSlice sync operation:

```
for each service S {
    all = accumulate numEndpoints(S)
    term = accumulate numTerminatingEndpoints(S)
    if numEndpoints(S) > 0 {
        ratio = accumulate  numTerminatingEndpoints(S) / numEndpoints(S)
    }
}
set a Gauge metric num-endpoints[50p, 90p, 99p] from "all"
set a Gauge metric num-terminating-endpoints[50p, 90p, 99p] from "term"
set a Gauge metric pct-terminating-endpoints[50p, 90p, 99p] from "ratio"
```

The percentiles are hardcoded here, but it seems super useful to know when `num-terminating-endpoints[90p] > 1` (for example)

Reply:
There may be a better way to express this - I just don't know prometheus metrics well enough.

Reply:
> That covers total endpoints across all EndpointSlices, that seems reasonable to me. Thoughts @wojtek-t ?

Yes - aggregations are definitely enough for me. Please do.

> I admit some ignorance of metrics compat - is it allowable to take a metric like EndpointsDesired which has no labels, and add new labels?

Yes - we're doing that pretty often.

> In this example wouldn't we end up double counting endpoints?

That's actually great point...
We would need to make them non-overlapping.


And going back to the question about usefulness of those metrics - I think that having visibility into them can be fairly useful "spot check" if something bad was happening (e.g. suddenly number of terminating suddenty grow up 2x or sth).

Reply:
re: double counting - yeah the sum of those is not useful :)  Maybe a better way to express it.

Reply:
> And going back to the question about usefulness of those metrics - I think that having visibility into them can be fairly useful "spot check" if something bad was happening (e.g. suddenly number of terminating suddenty grow up 2x or sth).

... or if you have enabled it and you don't have any Ready/Serving endpoint XD

Reply:
I'd like to know "how many times was traffic sent to a terminating/!ready endpoint because there was no other choice"

Reply:
> I'd like to know "how many times was traffic sent to a terminating/!ready endpoint because there was no other choice"

kube-proxy is not capable of knowing that level of detail, unless it starts inspecting packets. Maybe it can track # of connections through conntrack, but even that would add a lot of complexity in kube-proxy just to track this metric.